hs_core/views/resource_rest_api.py

```diff
@@ -893,6 +899,128 @@ def _validate_metadata(metadata_list):
             raise ValidationError(detail=err_message)
 
 
+def get_path(metadata):
+    hs_res_id = metadata.get('hs_res_id')
+    eventual_relative_path = ''
+    try:
+        # see if there is a path within data/contents that the file should be uploaded to
+        existing_path_in_resource = metadata.get('existing_path_in_resource', '')
+        existing_path_in_resource = json.loads(existing_path_in_resource).get("path")
+        if existing_path_in_resource:
+            # in this case, we are uploading to an existing folder in the resource
+            # existing_path_in_resource is a list of folder names
+            # append them into a path
+            for folder in existing_path_in_resource:
+                eventual_relative_path += folder + '/'
+    except Exception as ex:
+        logger.info(f"Existing path in resource not found: {str(ex)}")
+
+    # handle the case that a folder was uploaded instead of a single file
+    # use the metadata.relativePath to rebuild the folder structure
+    path_within_uploaded_folder = metadata.get('relativePath', '')
+    # path_within_resource_contents will include the name of the file, so we need to remove it
+    path_within_uploaded_folder = os.path.dirname(path_within_uploaded_folder)
+    if path_within_uploaded_folder:
+        eventual_relative_path += path_within_uploaded_folder + '/'
+    path = f'{hs_res_id}/data/contents/{eventual_relative_path}'
+    return path
+
+
+class CustomTusFile(TusFile):
+    # extend TusFile to allow it to write to s3 storage instead of local disk
+    # https://github.com/alican/django-tus/blob/2aac2e7c0e6bac79a1cb07721947a48d9cc40ec8/django_tus/tusfile.py#L52
```

Comment:
the bulk of this PR is extending django_tus to work with s3 instead of local disk